lib/Sema/TypeCheckType.cpp

```diff
@@ -319,6 +320,31 @@ Type TypeChecker::resolveTypeInContext(
         }
       }
     }
+    
+    // If we found an alias type in an inherited protocol, resolve it based on our
+    // own `Self`.
+    if (alias && alias->hasUnderlyingType() && alias->getUnderlyingType()->hasArchetype()) {
```

Comment:
We can validate the alias declaration rather than checking that it has an underlying archetype, right?